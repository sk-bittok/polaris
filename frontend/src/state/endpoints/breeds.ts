import type {
	RegisterBreedSchema,
	UpdateBreedSchema,
} from "@/lib/schemas/animal";
import type { Breed } from "@/models/livestock";
import type { EndpointBuilder } from "@reduxjs/toolkit/query";

export const breedEndpoints = (build: EndpointBuilder<any, any, any>) => ({
	getBreeds: build.query<Breed[], void>({
		query: () => "/breeds",
		providesTags: ["GetBreeds"],
	}),
	getBreedById: build.query<Breed, number>({
		query: (id) => ({ url: `/breeds/${id}` }),
		providesTags: (result, error, id) => [{ type: "GetBreeds", id }],
	}),
	createBreed: build.mutation<Breed, RegisterBreedSchema>({
		query: (params) => ({
			url: "/breeds",
			method: "POST",
			body: {
				specie: params.specie,
				name: params.name,
				description: params.description,
				typicalMaleWeightRange: params.maleWeightRange,
				typicalFemaleWeightRange: params.femaleWeightRange,
				typicalGestationPeriod: params.gestationPeriod,
			},
		}),
		invalidatesTags: ["GetBreeds"],
	}),
	deleteBreed: build.mutation<void, number>({
		query: (id) => ({
			url: `/breeds/${id}`,
			method: "DELETE",
		}),
		invalidatesTags: ["GetBreeds"],
	}),
	updateBreed: build.mutation<Breed, { data: UpdateBreedSchema; id: number }>({
		query: ({ data, id }) => ({
			url: `/breeds/${id}`,
			method: "PATCH",
			body: {
				specie: data.specie,
				name: data.name,
				description: data.description,
				typicalMaleWeightRange: data.maleWeightRange,
				typicalFemaleWeightRange: data.femaleWeightRange,
				typicalGestationPeriod: data.gestationPeriod,
			},
		}),
		invalidatesTags: (result, error, id) => [{ type: "GetBreeds", id }],
	}),
});
