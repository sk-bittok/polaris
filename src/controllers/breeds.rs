use axum::{
    Extension, Json, Router,
    body::Body,
    debug_handler,
    extract::{Path, Query, State},
    http::StatusCode,
    response::{IntoResponse, Response},
    routing::{delete, get, patch, post},
};
use serde_json::json;

use crate::{
    AppContext, Error, Result,
    middlewares::TokenClaims,
    models::{
        breeds::{Breed, BreedQuery},
        dto::{RegisterBreed, UpdateBreed},
        species::Specie,
        users::User,
    },
    views::animals::BreedResponse,
};

#[debug_handler]
async fn all(
    State(ctx): State<AppContext>,
    Extension(_auth): Extension<TokenClaims>,
    user: User,
    Query(conditions): Query<BreedQuery>,
) -> Result<Response> {
    let breeds = Breed::find_by_all(&ctx.db, user.organisation_pid, &conditions).await?;

    let breeds_futures = breeds
        .iter()
        .map(|breed| {
            let db = ctx.db.clone();
            let breed = breed.clone();
            async move {
                let specie = Specie::find_by_id(&db, breed.specie_id).await?;
                Ok::<_, Error>(BreedResponse::new(&breed, &specie.name))
            }
        })
        .collect::<Vec<_>>();

    let breeds_response = futures::future::try_join_all(breeds_futures).await?;

    Ok((StatusCode::OK, Json(breeds_response)).into_response())
}

#[debug_handler]
async fn one(State(ctx): State<AppContext>, user: User, Path(id): Path<i32>) -> Result<Response> {
    let breed = Breed::find_by_id(&ctx.db, id, user.organisation_pid).await?;
    let specie = Specie::find_by_id(&ctx.db, breed.specie_id).await?;

    Ok((
        StatusCode::OK,
        Json(BreedResponse::new(&breed, &specie.name)),
    )
        .into_response())
}

#[debug_handler]
async fn add(
    State(ctx): State<AppContext>,
    user: User,
    Json(params): Json<RegisterBreed<'static>>,
) -> Result<Response> {
    let mut tx = ctx.db.begin().await?;

    let specie = Specie::find_by_name(&mut *tx, &params.specie).await?;
    let breed = Breed::create(&mut *tx, user.organisation_pid, &params, specie.id).await?;

    tx.commit().await?;

    let response = Response::builder()
        .status(StatusCode::CREATED)
        .body(Body::new(
            json!(BreedResponse::new(&breed, &specie.name)).to_string(),
        ))?;

    Ok(response)
}

#[debug_handler]
async fn remove(
    State(ctx): State<AppContext>,
    user: User,
    Path(id): Path<i32>,
) -> Result<Response> {
    let mut tx = ctx.db.begin().await?;

    let _result = Breed::delete_breed(&mut *tx, user.organisation_pid, id).await?;

    tx.commit().await?;

    Ok((StatusCode::NO_CONTENT, Json(json!({}))).into_response())
}

#[debug_handler]
async fn update(
    State(ctx): State<AppContext>,
    user: User,
    Path(id): Path<i32>,
    Json(params): Json<UpdateBreed<'static>>,
) -> Result<Response> {
    let breed = Breed::update_by_id(&ctx.db, user.organisation_pid, id, &params).await?;

    Ok((StatusCode::CREATED, Json(breed)).into_response())
}

pub fn router(ctx: AppContext) -> Router {
    Router::new()
        .route("/", get(all))
        .route("/", post(add))
        .route("/{id}", get(one))
        .route("/{id}", delete(remove))
        .route("/{id}", patch(update))
        .with_state(ctx)
}
