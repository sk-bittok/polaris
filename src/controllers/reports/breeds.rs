use axum::{
    Json, Router, debug_handler,
    extract::{Query, State},
    http::StatusCode,
    response::{IntoResponse, Response},
    routing::{get, post},
};
use serde::{Deserialize, Serialize};

use crate::{
    AppContext, Result,
    models::{BreedSummary, BreedSummaryQuery, users::User},
};

#[derive(Debug, Deserialize, Clone, Serialize)]
struct GenerateReport {
    breed: String,
    specie: String,
}

#[debug_handler]
async fn add(
    State(ctx): State<AppContext>,
    user: User,
    Query(params): Query<GenerateReport>,
) -> Result<Response> {
    let generated = BreedSummary::generate(
        &ctx.db,
        user.organisation_pid,
        &params.breed,
        &params.specie,
    )
    .await?;

    Ok((StatusCode::CREATED, Json(generated)).into_response())
}

#[debug_handler]
async fn all(
    State(ctx): State<AppContext>,
    user: User,
    Query(params): Query<BreedSummaryQuery>,
) -> Result<Response> {
    let reports = BreedSummary::find_list(&ctx.db, user.organisation_pid, &params).await?;

    Ok((StatusCode::OK, Json(reports)).into_response())
}

pub fn router(ctx: AppContext) -> Router {
    Router::new()
        .route("/", post(add))
        .route("/", get(all))
        .with_state(ctx)
}
