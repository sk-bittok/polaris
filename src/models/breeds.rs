#![allow(clippy::missing_errors_doc)]
#![allow(clippy::missing_const_for_fn)]

use async_trait::async_trait;
use chrono::{DateTime, FixedOffset};
use serde::{Deserialize, Serialize};
use sqlx::{Encode, Executor, Postgres, postgres::PgQueryResult, prelude::FromRow};
use uuid::Uuid;

use crate::{models::species::Specie, seed::Seedable};

use super::{
    ModelError, ModelResult,
    dto::{RegisterBreed, UpdateBreed},
};

#[derive(Debug, Clone, Deserialize)]
pub struct BreedQuery {
    pub specie: Option<String>,
    pub name: Option<String>,
}

#[derive(Debug, Deserialize, Serialize, Clone, FromRow, Encode)]
pub struct Breed {
    pub(crate) id: i32,
    pub(crate) specie_id: i32,
    pub(crate) name: String,
    pub(crate) description: Option<String>,
    pub(crate) typical_male_weight_range: Option<String>,
    pub(crate) typical_female_weight_range: Option<String>,
    pub(crate) typical_gestation_period: Option<String>,
    pub(crate) created_at: DateTime<FixedOffset>,
    pub(crate) updated_at: DateTime<FixedOffset>,
    pub(crate) organisation_pid: Option<Uuid>,
    pub(crate) is_system_defined: bool,
}

impl Breed {
    pub async fn find_by_id<'e, C>(db: C, id: i32, org_pid: Uuid) -> ModelResult<Self>
    where
        C: Executor<'e, Database = Postgres>,
    {
        sqlx::query_as::<_, Self>("SELECT * FROM breeds WHERE id = $1 AND (is_system_defined = TRUE OR organisation_pid = $2)")
            .bind(id)
            .bind(org_pid)
            .fetch_optional(db)
            .await?
            .ok_or_else(|| ModelError::EntityNotFound)
    }

    pub async fn find_by_all<'e, C>(
        db: C,
        org_pid: Uuid,
        conditions: &BreedQuery,
    ) -> ModelResult<Vec<Self>>
    where
        C: Executor<'e, Database = Postgres>,
    {
        let mut query =
            sqlx::query_as::<_, Self>("SELECT * FROM breeds WHERE  (is_system_defined = TRUE OR organisation_pid = $1) ORDER BY name")
                .bind(org_pid);

        if let Some(name) = &conditions.name {
            query = sqlx::query_as::<_, Self>(
                "SELECT * FROM breeds WHERE (is_system_defined = TRUE OR organisation_pid = $1) AND name LIKE $2",
            )
            .bind(org_pid)
            .bind(format!("%{name}%"));
        }

        if let Some(specie) = &conditions.specie {
            query = sqlx::query_as::<_, Self>(
                "SELECT * FROM breeds WHERE (is_system_defined = TRUE OR organisation_pid = $1 ) AND specie LIKE $2",
            )
            .bind(org_pid)
            .bind(format!("%{specie}%",));
        }

        query.fetch_all(db).await.map_err(Into::into)
    }

    pub async fn create<'e, C>(
        db: C,
        org_pid: Uuid,
        dto: &RegisterBreed<'_>,
        specie_id: i32,
    ) -> ModelResult<Self>
    where
        C: Executor<'e, Database = Postgres>,
    {
        let query = sqlx::query_as::<_, Self>(
            "INSERT INTO breeds (specie_id, name, typical_male_weight_range, typical_female_weight_range, typical_gestation_period, organisation_pid, description)
                VALUES ($1, $2, $3, $4, $5, $6, $7) RETURNING *
            ")
        .bind(specie_id)
        .bind(&dto.name)
        .bind(&dto.typical_male_weight_range)
        .bind(&dto.typical_female_weight_range)
        .bind(&dto.typical_gestation_period)
        .bind(org_pid)
        .bind(&dto.description)
        .fetch_one(db)
        .await?;

        Ok(query)
    }

    pub async fn delete_breed<'e, C>(db: C, org_pid: Uuid, id: i32) -> ModelResult<PgQueryResult>
    where
        C: Executor<'e, Database = Postgres>,
    {
        let query = sqlx::query("DELETE FROM breeds WHERE organisation_pid = $1 AND id = $2")
            .bind(org_pid)
            .bind(id)
            .execute(db)
            .await?;

        Ok(query)
    }

    pub async fn seed(db: &sqlx::PgPool, path: &str) -> ModelResult<()> {
        let breeds = Self::load_file(path).await?;

        Self::seed_data(db, &breeds).await
    }

    pub async fn update_by_id<'e, C>(
        db: &C,
        org_pid: Uuid,
        id: i32,
        params: &UpdateBreed<'_>,
    ) -> ModelResult<Self>
    where
        for<'a> &'a C: Executor<'e, Database = Postgres>,
    {
        let model = Self::find_by_id(db, id, org_pid).await?;

        let name = params
            .name
            .as_ref()
            .map_or_else(|| model.name.as_str().to_string(), ToString::to_string);
        let description = params
            .description
            .as_ref()
            .map_or_else(|| model.description, |desc| Some(desc.to_string()));
        let male_weight = params.typical_male_weight_range.as_ref().map_or_else(
            || model.typical_male_weight_range,
            |weight| Some(weight.to_string()),
        );
        let female_weight = params.typical_female_weight_range.as_ref().map_or_else(
            || model.typical_female_weight_range,
            |weight| Some(weight.to_string()),
        );
        let gestation_period = params.typical_gestation_period.as_ref().map_or_else(
            || model.typical_gestation_period,
            |period| Some(period.to_string()),
        );

        let specie = if let Some(specie) = params.specie.as_ref() {
            Specie::find_by_name(db, specie).await?
        } else {
            Specie::find_by_id(db, model.specie_id).await?
        };

        let query = sqlx::query_as::<_, Self>(
            "
                UPDATE breeds
                SET specie_id = $3,
                name = $4, description = $5,
                typical_male_weight_range = $6,
                typical_female_weight_range = $7,
                typical_gestation_period = $8
                WHERE id = $1 AND organisation_pid = $2
                RETURNING *
            ",
        )
        .bind(id)
        .bind(org_pid)
        .bind(specie.id())
        .bind(&name)
        .bind(&description)
        .bind(&male_weight)
        .bind(&female_weight)
        .bind(&gestation_period)
        .fetch_one(db)
        .await?;

        Ok(query)
    }
}

#[async_trait]
impl Seedable for Breed {
    async fn seed_data(db: &sqlx::PgPool, breeds: &[Self]) -> ModelResult<()> {
        for breed in breeds {
            sqlx::query("INSERT INTO breeds 
            (id, organisation_pid, specie_id, name, description, is_system_defined, typical_male_weight_range,
            typical_female_weight_range, typical_gestation_period, created_at, updated_at )
            VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11 )")
            .bind(breed.id)
            .bind(breed.organisation_pid)
            .bind(breed.specie_id)
            .bind(breed.name.as_str())
            .bind(breed.description.as_deref())
            .bind(breed.is_system_defined)
            .bind(breed.typical_male_weight_range.as_deref())
            .bind(breed.typical_female_weight_range.as_deref())
            .bind(breed.typical_gestation_period.as_deref())
            .bind(breed.created_at)
            .bind(breed.updated_at)
            .execute(db)
            .await?;
        }

        Ok(())
    }
}
