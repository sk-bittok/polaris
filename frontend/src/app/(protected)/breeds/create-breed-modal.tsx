import CustomFormField from "@/components/form-field";
import { Form } from "@/components/ui/form";
import { registerBreedSchema, RegisterBreedSchema, Category } from "@/models/livestock";
import { zodResolver } from "@hookform/resolvers/zod";
import { useState } from "react";
import { useForm } from "react-hook-form";
import { DialogHeader, DialogTitle, DialogDescription, DialogContent, Dialog, DialogClose, DialogTrigger } from "@/components/ui/dialog";

const speciesOptions = [
  { label: "Cattle", value: Category.Cattle },
  { label: "Sheep", value: Category.Sheep },
  { label: "Goat", value: Category.Goat },
  { label: "Pig", value: Category.Pig },
  { label: "Chicken", value: Category.Chicken },
];

type Props = {
  isOpen: boolean;
  onClose: () => void,
  onCreate: (data: RegisterBreedSchema) => void,
  children: React.ReactNode,
}

export default function CreateBreedModal({ isOpen, onClose, onCreate, children }: Props) {
  const [formData, setFormData] = useState<RegisterBreedSchema>({
    specie: "",
    name: "",
    description: "",
    maleWeightRange: "",
    femaleWeightRange: "",
    gestationPeriod: "",
  });

  const form = useForm<RegisterBreedSchema>({
    resolver: zodResolver(registerBreedSchema),
    mode: 'onChange',
    defaultValues: formData
  });

  const handleSubmit = (data: RegisterBreedSchema) => {
    onCreate(data);
    form.reset();
    onClose();
  };

  const handleClose = () => {
    form.reset();
    onClose();
  };


  return (
    <Dialog open={isOpen} onOpenChange={(open) => {
      if (!open) {
        handleClose();
      }
    }}>
      <DialogTrigger asChild>{children}</DialogTrigger>
      <DialogContent>
        <DialogHeader>
          <DialogTitle>Register new breed</DialogTitle>
          <DialogDescription>Register a new breed. Click save when your are done</DialogDescription>
        </DialogHeader>

        {/* Scrollabel form area */}
        <div className="overflow-y-auto p-6 max-h-[calc(90vh-130px)]">
          <Form {...form}>
            <form onSubmit={form.handleSubmit(handleSubmit)} className="">
              <div className="space-y-5">

                {/* First column */}
                <div className="space-y-5">
                  <CustomFormField
                    control={form.control}
                    name="specie"
                    type="select"
                    options={speciesOptions}
                    placeholder="Select a specie"
                    label="Species"
                    inputClassName="w-full"
                  />

                  <CustomFormField
                    control={form.control}
                    name="name"
                    placeholder="i.e. Hereford, Dorper"
                    label="Name"
                  />
                  <div className="grid grid-cols-1 md:grid-cols-2 gap-5">
                    <CustomFormField
                      control={form.control}
                      name="maleWeightRange"
                      placeholder="e.g., 1000-1200 kg"
                      label="Male Weight Range"
                    />

                    {/* Second column */}
                    <CustomFormField
                      control={form.control}
                      name="femaleWeightRange"
                      placeholder="e.g., 550-750 kg"
                      label="Female Weight Range"
                    />
                  </div>

                  <CustomFormField
                    control={form.control}
                    name="gestationPeriod"
                    placeholder="e.g., 278 days"
                    label="Gestation Period"
                  />

                  {/* Full width description field */}
                  <CustomFormField
                    control={form.control}
                    name="description"
                    type="textarea"
                    placeholder="Describe the breed's characteristics, history, purpose, and notable features..."
                    label="Description"
                    inputClassName="min-h-32"
                  />
                </div>
              </div>

              <div className="p-6 border-t border-gray-200 dark:border-gray-700 bg-gray-50 dark:bg-gray-900 mt-2 rounded-lg shadow-md flex items-center justify-end gap-3">
                <button
                  type="button"
                  onClick={handleClose}
                  className="px-4 py-2 border border-gray-300 dark:border-gray-600 rounded-md text-sm font-medium text-gray-700 dark:text-gray-300 hover:bg-gray-50 dark:hover:bg-gray-700 transition-colors focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-blue-500"
                >Cancel
                </button>

                <button
                  type="submit" className="flex items-center px-4 py-2 bg-green-600 text-white rounded-md text-sm font-medium hover:bg-green-700 transition-colors focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-green-500 disabled:opacity-50 disabled:hover:bg-green-600 disabled:cursor-not-allowed"
                >
                  Create Breed
                </button>
              </div>
            </form>
          </Form>
        </div>
      </DialogContent>
    </Dialog>
  );
}

