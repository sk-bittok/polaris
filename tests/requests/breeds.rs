use axum::http::StatusCode;
use insta::{Settings, assert_debug_snapshot, with_settings};
use polaris::models::breeds::Breed;
use serial_test::serial;

use crate::request;

use crate::requests::prepare_auth;

macro_rules! configure_insta {
    ($(expr:expr),*) => {
        let mut settings = Settings::clone_current();
        settings.set_prepend_module_to_snapshot(false);
        settings.set_snapshot_path("snapshots/breeds");
        settings.set_snapshot_suffix("breeds");
        let _guard = settings.bind_to_scope();
    };
}

#[tokio::test]
#[serial]
async fn can_fetch_all() {
    request(|server, context| async move {
        configure_insta!();

        crate::seed_data(&context.db)
            .await
            .inspect_err(|e| eprintln!("Seed error {:?}", e))
            .unwrap();

        let user = prepare_auth::init_login(&server, &context).await;
        let (auth_header, auth_value) = prepare_auth::auth_header(user.access_token);

        let request = server
            .get("/breeds")
            .add_header(auth_header, auth_value)
            .await;

        with_settings!({
            filters => {
                let mut filters = crate::cleanup_date().to_vec();
                filters.extend(crate::cleanup_uuid().to_vec());
                filters.extend(crate::cleanup_int().to_vec());
                filters
            }
        }, {
            assert_debug_snapshot!((request.status_code(), request.text()));
        });
    })
    .await;
}

#[tokio::test]
#[serial]
async fn can_fetch_by_id() {
    request(|server, context| async move {
        configure_insta!();

        let user = prepare_auth::init_login(&server, &context).await;
        let (auth_header, auth_value) = prepare_auth::auth_header(user.access_token);

        let request = server
            .get("/breeds/1")
            .add_header(auth_header, auth_value)
            .await;

        with_settings!({
            filters => {
                let mut filters = crate::cleanup_date().to_vec();
                filters.extend(crate::cleanup_uuid().to_vec());
                filters.extend(crate::cleanup_int().to_vec());
                filters
            }
        }, {
            assert_debug_snapshot!((request.status_code(), request.text()));
        });
    })
    .await;
}

#[tokio::test]
#[serial]
async fn can_create_one() {
    request(|server, context| async move {
        configure_insta!();

        let post_data = serde_json::json!({
            "name": "Fleckvieh",
            "description": "A dual purpose (milk and meat) breed of cattle with a white head and either a brown, red or yellow with white sections body",
            "typicalGestationPeriod": "285 days",
            "specie": "cattle",
            "typicalMaleWeightRange": "1000-1300",
            "typicalFemaleWeightRange": "600-800"
        });

        let user = prepare_auth::init_login(&server, &context).await;

        let (auth_header, auth_value) = prepare_auth::auth_header(user.access_token);

        let request = server.post("/breeds")
            .json(&post_data)
            .add_header(auth_header, auth_value).await;

        assert_eq!(request.status_code(), StatusCode::CREATED);
    })
    .await;
}

#[tokio::test]
#[serial]
async fn can_delete_by_id() {
    request(|server, context| async move {
        configure_insta!();

        crate::seed_data(&context.db).await.unwrap();

        let post_data = serde_json::json!({
            "name": "Fleckvieh",
            "description": "A dual purpose (milk and meat) breed of cattle with a white head and either a brown, red or yellow with white sections body",
            "typicalGestationPeriod": "285 days",
            "specie": "cattle",
            "typicalMaleWeightRange": "1000-1300 kg",
            "typicalFemalWeightRange": "600-800 kg"
        });
        let user = prepare_auth::init_login(&server, &context).await;
        let (auth_header, auth_value) = prepare_auth::auth_header(user.access_token);
        server
            .post("/breeds")
            .json(&post_data)
            .add_header(&auth_header, &auth_value)
            .await;

        let request = server
            .delete("/breeds/8")
            .add_header(auth_header, auth_value)
            .await;

        assert_eq!(request.status_code(), StatusCode::NO_CONTENT);
    }).await;
}

#[tokio::test]
#[serial]
async fn can_update_by_id() {
    request(|server, context| async move {
        configure_insta!();

        crate::seed_data(&context.db).await.unwrap();

        let post_data = serde_json::json!({
            "name": "Fleckvieh",
            "description": "A dual purpose (milk and meat) breed of cattle with a white head and either a brown, red or yellow with white sections body",
            "typicalGestationPeriod": "285 days",
            "specie": "cattle",
            "typicalMaleWeightRange": "1000-1300 kg",
            "typicalFemalWeightRange": "600-800 kg"
        });
        let user = prepare_auth::login_user(&server, &context).await;
        let (auth_header, auth_value) = prepare_auth::auth_header(user.access_token);
        let _post_request = server
            .post("/breeds")
            .json(&post_data)
            .add_header(&auth_header, &auth_value)
            .await;

        let patch_data = serde_json::json!({
           "description": "Recently, great gains have been made in the genetic milk productiveness of the Fleckvieh through breed management to the point that the 
           Fleckvieh rivals pure milk producing dairy breeds. Special value is placed on the Fleckvieh’s fitness characteristics including fertility, longevity, calving ease, udder health, milking speed, somatic cell count, and persistence. A large percentage of young Fleckvieh cattle are on alpine grazing which results in good overall health and longevity of life. Breeding target: 38 per cent milk, 16 per cent beef, 46 per cent fitness. Fleckvieh cattle exhibit good development and performance capacity along with good conformation. 6,000 kg milk in the 1st lactation and over 7,000 – 9,000 kg milk in later lactations with 4.2 per cent fat and 3.7 per cent protein.
        Beef performance:
            Daily gain: 1.44 kg
            Carcass weight: 57.2 per cent
            EUROP classification E and U: > 85.7 per cent",
            "typicalFemaleWeight": "650-850 kg", 
        });

        let response = server.patch("/breeds/10").json(&patch_data).add_header(&auth_header, &auth_value).await;

        assert_eq!(response.status_code(), StatusCode::CREATED);

        with_settings!({
            filters => {
                let mut filters = crate::cleanup_date().to_vec();
                filters.extend(crate::cleanup_int().to_vec());
                filters
            }
        }, {
            assert_debug_snapshot!((response.status_code(), response.json::<Breed>()));    
        });

    }).await;
}
