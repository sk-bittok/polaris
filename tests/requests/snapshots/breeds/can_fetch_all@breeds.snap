---
source: tests/requests/breeds.rs
assertion_line: 46
expression: "(request.status_code(), request.text())"
---
(
    200,
    "[{\"id\":1,\"organisationPid\":null,\"specie\":\"cattle\",\"name\":\"Aberdeen Angus\",\"maleWeightRange\":\"750-950 kg\",\"femaleWeightRange\":\"500-550 kg\",\"gestationPeriod\":\"283 days\",\"description\":\"Aberdeen Angus cattle are naturally polled and can be black or red in colour although black is the dominant colour, white may occasionally appear on the udder. They are resistant to harsh weather, undemanding, adaptable, good natured, mature extremely early and have a high carcass yield with nicely marbled meat. Angus are renowned as a carcass breed. They are used widely in crossbreeding to improve carcass quality and milking ability. Angus females calve easily and have good calf rearing ability. They are also used as a genetic dehorner as the polled gene is passed on as a dominant characteristic.\",\"createdAt\":\"DATE\",\"isSystemDefined\":true},{\"id\":3,\"organisationPid\":null,\"specie\":\"cattle\",\"name\":\"Beef Shorthorn\",\"maleWeightRange\":\"1100-1300 kg\",\"femaleWeightRange\":\"600-800 kg\",\"gestationPeriod\":\"284 days\",\"description\":\"Beef Shorthorn come in three colours, red, white and roan. Red cattle may be solid red or have white markings and they can be horned or polled. They are bigger than their dairy counterparts and are grown specifically for their beef. Solid red Shorthorn are often used to help maintain solid colors in crosses. The whites and roans can be used on black cattle to get both blues and blacks. Shorthorn bulls are active aggressive breeders and we get many reports of bulls being used in commercial herds for several years, often up to eight years. An advantage of the Shorthorn cross is that the steers produced have an excellent rate of gain, good feed conversion and increased marbling and tenderness. Overall the Shorthorn is the ideal breed for the production of a choice high quality beef with its suitability for extensive and organic farming systems and its proven marbling and early finishing abilities.\",\"createdAt\":\"DATE\",\"isSystemDefined\":true},{\"id\":5,\"organisationPid\":null,\"specie\":\"cattle\",\"name\":\"Boran\",\"maleWeightRange\":\"700-900 kg\",\"femaleWeightRange\":\"450-585 kg\",\"gestationPeriod\":\"285 days\",\"description\":\"The Boran is medium in size with a short head, small ears, loose dewlap and a large hump above the shoulders. They can be horned or polled. They vary in height from 114cm to 147cm tall, and in weight bulls weigh approximately 500kg to 850kg. Cows weigh about 380kg to 450kg Their skin is loose, thick and extremely pliable for added insect repellence plus it is dark pigmented with fine short hair for heat tolerance. Hair colour can be a range of colours except brindle or solid black. The Boran male and female share breed points, the sexes, however, show marked dimorphism - the female being notably small, whilst the male grows to a large size. The cow has a well-carried udder with strong attachments and neat, small teats, in contrast to some Asian Zebu breeds. Boran heifers reach puberty at an average age of 385 days. She is an excellent mother, not only will she feed her calf so well that high weaning weights are attainable, but she guards against predators, and will never allow her calf to get lost in the bush. Calving problems hardly exist. Calves at birth weigh an average of 28 kg for males and females, 25 kg. Boran cattle have developed adaptive traits of crucial importance for their survival. Some of these characters are - the ability to withstand periodic shortage of water and feed, ability to walk long distances in search of water and feed and ability to digest low quality feeds. The herd instinct of the Boran makes it easy to manage and survive in bush country. They will always stay together and can graze on the trot. The well-developed beef conformation shows up in carcase appraisals. The depth of eye muscle, marbling, even fat cover and ratio of hind to forequarter make the Boran difficult to beat, hence the preference of Kenya butchers for young, well-finished Boran steers.\",\"createdAt\":\"DATE\",\"isSystemDefined\":true},{\"id\":7,\"organisationPid\":null,\"specie\":\"sheep\",\"name\":\"Dorper\",\"maleWeightRange\":\"50-80\",\"femaleWeightRange\":\"90-140\",\"gestationPeriod\":\"147 days\",\"description\":\"Mutton sheep with black head and white coat\",\"createdAt\":\"DATE\",\"isSystemDefined\":true},{\"id\":8,\"organisationPid\":null,\"specie\":\"sheep\",\"name\":\"Dorset Horn\",\"maleWeightRange\":\"60-75\",\"femaleWeightRange\":\"90-120\",\"gestationPeriod\":\"145 days\",\"description\":\"White mutton sheep from England\",\"createdAt\":\"DATE\",\"isSystemDefined\":true},{\"id\":4,\"organisationPid\":null,\"specie\":\"cattle\",\"name\":\"Friesian\",\"maleWeightRange\":\"1000-1200 kg\",\"femaleWeightRange\":\"580-750 kg\",\"gestationPeriod\":\"279 days\",\"description\":\"The Friesian can be one of two coat colour types, white with black patches (the common colour) or white with red patches. They are very similar in size and confirmation to the Holstein. The Friesian is a renowned dairy breed with some outstanding examples of the breed having 12 to 15 lactations to their credit, emphasising their inherent natural fecundity. In response to demand, protein percentages have been raised across the breed and herd protein levels of 3.4% to 3.5% are not uncommon. One of the great strengths of the British Friesian is the ability of the male calf to finish and grade satisfactorily, either in intensive systems, or as steers, extensively.\",\"createdAt\":\"DATE\",\"isSystemDefined\":true},{\"id\":2,\"organisationPid\":null,\"specie\":\"cattle\",\"name\":\"Hereford\",\"maleWeightRange\":\"1000-1045 kg\",\"femaleWeightRange\":\"545-680 kg\",\"gestationPeriod\":\"285 days\",\"description\":\"The modern Hereford is coloured dark red to red-yellow, with a white face, crest, dewlap, and underline. Herefords with white flanks and white markings below the knees and hocks are also common. Most animals have short thick horns that typically curve down at the sides of the head, but there is a polled strain in North America and UK (Polled Hereford). Mature males may weigh up to 1000 kg, while mature females may weigh around 600 kg. They are muscular, moderate to long in length of side, adequate in length of leg, large in size, trim, and smooth. They are also well developed in the regions of valuable cuts - the back, loin, and hind quarters or round. These cattle are known for their vigor and foraging ability and for their longevity, many females live and produce calves beyond the age of 15 years. Bulls are capable of remaining profitable at stud to the age of 12 or more. Many breeders keep their elderly cattle until they die of natural causes. Herefords will stand out in the arctic snows of Finland, endure the heat of Northern Transvaal, withstand the tough climate and rough grazing of northern Uruguay or the sub-tropical zones of Brazil and continue to thrive. Herefords are generally docile and fast growing cattle with good beef quality.\",\"createdAt\":\"DATE\",\"isSystemDefined\":true},{\"id\":9,\"organisationPid\":null,\"specie\":\"sheep\",\"name\":\"Merino\",\"maleWeightRange\":\"30-60\",\"femaleWeightRange\":\"70-80\",\"gestationPeriod\":\"144 days\",\"description\":\"Wool sheep\",\"createdAt\":\"DATE\",\"isSystemDefined\":true},{\"id\":6,\"organisationPid\":null,\"specie\":\"cattle\",\"name\":\"Sahiwal\",\"maleWeightRange\":\"650-850 kg\",\"femaleWeightRange\":\"350-550 kg\",\"gestationPeriod\":\"289 days\",\"description\":\"Their colour can range from reddish brown through to the more predominant red, with varying amounts of white on the neck, and the underline. In males the colour darkens towards the extremities, such as the head, legs and tail. It is tick-resistant, heat-tolerant and noted for its high resistance to parasites, both internal and external. Cows average 2270kg of milk during a lactation while suckling a calf and much higher milk yields have been recorded. They are generally docile and lethargic, making them more useful for slow work. The Sahiwal is the heaviest milker of all Zebu breeds and display a well developed udder. Sahiwals demonstrate the ability to sire small, fast-growing calves and are noted for their hardiness under unfavorable climatic conditions.\",\"createdAt\":\"DATE\",\"isSystemDefined\":true}]",
)
